amaranth/hdl/ast.py

```diff
@@ -239,29 +263,38 @@ def __rrshift__(self, other):
         self.__check_shamt()
         return Operator(">>", [other, self])
 
+    @overridable_by_reflected("__rand__")
     def __and__(self, other):
         return Operator("&", [self, other])
     def __rand__(self, other):
         return Operator("&", [other, self])
+    @overridable_by_reflected("__rxor__")
     def __xor__(self, other):
         return Operator("^", [self, other])
     def __rxor__(self, other):
         return Operator("^", [other, self])
+    @overridable_by_reflected("__ror__")
     def __or__(self, other):
         return Operator("|", [self, other])
     def __ror__(self, other):
         return Operator("|", [other, self])
 
+    @overridable_by_reflected("__eq__")
     def __eq__(self, other):
         return Operator("==", [self, other])
+    @overridable_by_reflected("__ne__")
     def __ne__(self, other):
         return Operator("!=", [self, other])
+    @overridable_by_reflected("__ge__")
```

Comment:
I think this should be `__gt__` (and similar for the next 3 operators as well

Reply:
You're right, I had a brainfart and was thinking inverted result, not reflected operation. Thanks!

Reply:
I don't think this is right. Look up the `NotImplemented` value and its semantics.

Reply:
The intention here is to get behavior consistent with how Python treats derived classes that overrides the operators. The derived class' implementation will be tried first and if it returns `NotImplemented`, the parent's implementation will be tried.

Example:
```python
class Left:
    def __mul__(self, other):
        print("Left.__mul__ called")

    def __rmul__(self, other):
        print("Left.__rmul__ called")

class Right(Left):
    def __mul__(self, other):
        print("Right.__mul__ called")
        return NotImplemented

    def __rmul__(self, other):
        print("Right.__rmul__ called")
        return NotImplemented

Left() * Right()

'''
$ ./test_ops.py 
Right.__rmul__ called
Left.__mul__ called
'''
```

Reply:
So why can we not use the built-in behavior? I'm concerned about implementing a chunk of what Python should be handling itself, and its correctness.

Reply:
Oh, no, I misunderstood what you're doing. I think this is fine.

Reply:
Because `ValueCastable` is not a subclass of `Value`, and when the types are unrelated, the left operand will be tried first:
```python
class Left:
    def __mul__(self, other):
        print("Left.__mul__ called")

    def __rmul__(self, other):
        print("Left.__rmul__ called")

class Right:
    def __mul__(self, other):
        print("Right.__mul__ called")
        return NotImplemented

    def __rmul__(self, other):
        print("Right.__rmul__ called")
        return NotImplemented

Left() * Right()

'''
$ ./test_ops.py 
Left.__mul__ called
'''
```

Reply:
Yep, you're right. I was misremembering how the mechanism works. Apologies!